xarray/coding/variables.py

```diff
@@ -273,9 +282,19 @@ def decode(self, variable: Variable, name: T_Name = None):
                     SerializationWarning,
                     stacklevel=3,
                 )
-
-            dtype, decoded_fill_value = dtypes.maybe_promote(data.dtype)
-
+            units = attrs.get("units", None)
```

Comment:
Can we move `_FillValue` to `encoding` in

https://github.com/pydata/xarray/blob/51554f2638bc9e4a527492136fe6f54584ffa75d/xarray/coding/times.py#L723

 and then `CFMaskCoder` would not do anything?

Reply:
That should work, but we might have to do the casting to final dtype (which is done in `NonStringCoder`) in `CFDatetimeCoder` too. 

I'll have a go on this suggestion, thanks @dcherian.

Reply:
but at decode, the final dtype is always `datetime64[ns]`?

Reply:
In most cases, but it can also be an array of cftime-objects. I've once again restructured, but I'm not sure if this fit's all cases.